package uk.ac.ebi.protvar.utils;

public class Constants {
	public static final String GENOMIC_SYMBOL = ":g.";
	public static final String HGNC = "HGNC";
	public static final String SLASH = "/";
	public static final String SPLITTER = "-";
	public static final String VARIANT_SEPARATOR = ">";
	public static final String FILE_PATH = "./";
	public static final String SPACE = " ";
	public static final String COMMA = ",";

	public static final String NA = "N/A";
	public static final String NOTE_MAPPING_NOT_FOUND = "Mapping not found.";
	public static final String NOTE_INVALID_INPUT_CHROMOSOME = "Invalid chromosome ";
	public static final String NOTE_INVALID_INPUT_POSITION = "Position should be a number ";
	public static final String NOTE_INVALID_INPUT_ALT = "Invalid alternative ";
	public static final String NOTE_INVALID_INPUT_REF = "Invalid reference ";
	public static final String NOTE_INVALID_INPUT_FORMAT = "Invalid input format ";
	public static final String NOTE_INVALID_INPUT_NON_HUMAN_ACC = "Not a human UniProt reviewed protein ";
	public static final String NOTE_INVALID_INPUT_ACCESSION = "Invalid accession ";
	public static final String NOTE_INVALID_INPUT_PROTEIN_POSITION = "Invalid protein position ";
	public static final String NOTE_INVALID_INPUT_AA = "Invalid reference/alternative amino acid ";
	public static final String NOTE_INVALID_INPUT_NON_SNV = "Non SNV ";
}
